KFSPTS-28268 implement jaggaer upload suppliers XML step

jhulslander opened the pull request.

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,225 @@
+    }
+    
+    private SupplierResponseMessage buildSupplierResponseMessage(String responseString) {
+        SupplierSyncMessage synchMessage;
```
Should this be named `syncMessage` instead for consistency?

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,225 @@
+            
+            disableRequestChunkingIfNecessary(client, requestBuilder);
+            
+            Response response = requestBuilder.accept(MediaType.APPLICATION_XML)
```
You may want to close the `Response` object when you're done with it. The `CURestClientUtils` class has a convenience method for closing it quietly, if it's not possible or practical to use it in a try-with-resources block.

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,225 @@
+        }
+    }
+    
+    private void waitBetweenTries() {
```
Is this wait needed? Our calls to the new Concur APIs currently don't insert a wait/delay between retries.

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,225 @@
+    
+    private void generateResultsReport(List<JaggaerUploadFileResultsDTO> resultsList) {
+        /*
+         * this will be implemented as a report on Jira 
```
Do you need this comment? If so, I would suggest updating it to mention the specific JIRA ticket that will be implementing the report.

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -121,6 +126,12 @@ public <T> T unmarshalString(String xmlString, Class<T> clazz) throws JAXBExcept
+    
+    @Override
+    public <T> T unmarshalStringIgnoreDtd(String xmlString, Class<T> clazz) throws JAXBException, XMLStreamException, IOException {
+        InputStream axmlStream = IOUtils.toInputStream(xmlString);
```
Minor typo in variable name.

Also, is there a variant of `toInputStream()` that allows you to set the encoding?

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -141,6 +152,19 @@ public <T> T unmarshalStream(InputStream inputStream, Class<T> clazz)
+
+        XMLInputFactory xif = XMLInputFactory.newFactory();
+        xif.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+        XMLStreamReader xsr = xif.createXMLStreamReader(inputStream);
```
You may want to close the reader when you're done with it. The `CuXMLStreamUtils` class has a convenience method for closing it quietly.

Also, if the caller is supposed to be responsible for closing the input stream, then please wrap the stream argument in a `CloseShieldInputStream` before creating the reader. There should be an example in one of this class's other methods.

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/DisposableClientServiceImplBase.java:
```diff
@@ -55,5 +63,19 @@ protected void closeClientQuietly() {
         CURestClientUtils.closeQuietly(jerseyClient);
     }
+    
+    protected void disableRequestChunkingIfNecessary(Client client, Invocation.Builder requestBuilder) {
```
Now that you've moved this method to the superclass, the `client` arg is masking the `client` instance variable. You may want to make adjustments to avoid confusion, such as renaming the arg or making the method a static one.

cah292 reviewed (changes requested):
Here are the initial items I noticed when reviewing this.

Also, if it ends up not being practical to write a test that utilizes the JAX-RS client calls, you may want to create a follow-up ticket to do so. (The related PaymentWorks example is fairly complex because its test uses HTTP libraries; we have some other examples that use Spring MVC but I don't think the latter ones have multipart request examples.)

jhulslander commented:
Chad,
This is NOT ready for review again,  these changes aren't working yet and there is debug code in place.  However, I could use some help.  When I try to run the upload step I get the error shown below.  I am not sure what configuration I should be using to use WebClient.  I added spring-webflux using our current spring version to the POM file and also added reactor-netty-http per https://stackoverflow.com/questions/62933054/cannot-construct-spring-webclient-due-to-java-lang-illegalstateexception-no-sui .  Do you have any suggestions>=?

`2023-07-28 16:00:50,799 [KSB-Scheduled-pool-2-thread-1] u:/d: ERROR edu.cornell.kfs.module.purap.batch.service.impl.JaggaerUploadFileServiceImpl :: JaggaerUploadFileResultsDTO, got an exception calling the Jaggaer Web Service
java.lang.NoClassDefFoundError: io/netty/handler/codec/DefaultHeaders$ValueValidator
	at reactor.netty.http.client.HttpClientConfig.<init>(HttpClientConfig.java:350) ~[reactor-netty-http-1.1.9.jar:1.1.9]
	at reactor.netty.http.client.HttpClientConnect.<init>(HttpClientConnect.java:86) ~[reactor-netty-http-1.1.9.jar:1.1.9]
	at reactor.netty.http.client.HttpClient.create(HttpClient.java:398) ~[reactor-netty-http-1.1.9.jar:1.1.9]
	at org.springframework.http.client.reactive.ReactorClientHttpConnector.<init>(ReactorClientHttpConnector.java:56) ~[spring-web-5.3.27.jar:5.3.27]
	at org.springframework.web.reactive.function.client.DefaultWebClientBuilder.initConnector(DefaultWebClientBuilder.java:289) ~[spring-webflux-5.3.27.jar:5.3.27]
	at org.springframework.web.reactive.function.client.DefaultWebClientBuilder.build(DefaultWebClientBuilder.java:266) ~[spring-webflux-5.3.27.jar:5.3.27]
	at org.springframework.web.reactive.function.client.WebClient.create(WebClient.java:144) ~[spring-webflux-5.3.27.jar:5.3.27]
	at edu.cornell.kfs.module.purap.batch.service.impl.JaggaerUploadFileServiceImpl.uploadJaggaerXmlFile(JaggaerUploadFileServiceImpl.java:86) ~`

cah292 commented:
@jhulslander It looks like in the 2023-03-15 financials patch, KualiCo added a `WebClientFactory` helper class to assist with creating `WebClient` instances. You could try backporting that class or adding a CU-specific variant of it.

https://github.com/KualiCo/financials/blob/master/kfs-core/src/main/java/org/kuali/kfs/sys/web/WebClientFactory.java

Note that KualiCo configured the factory so that the web clients will follow redirects. If that's not the behavior you want, then you should create your own variant.

If you do create your own custom initialization, it seems that the key part is setting up an `HttpComponentsClientHttpConnector` instance on the client. That will make the client use the Apache HTTP Components/Client APIs (which base financials still uses).

jhulslander commented:
@cah292 ,
I was able to backport WebClientFactory but still am having problems.  I get the error shown below. Our current version of httpclient and httpcore are the versions used in base code when WebClientFactory was added, so I don't think that is the problem.  I thought I might have been using post wrong, so I tried using a couple of examples I found on stack over flow, all gave me versions of the error below.  I am not sure what is going awry.

`2023-08-01 16:13:21,570 [KSB-Scheduled-pool-2-thread-1] u:/d: WARN  reactor.core.Exceptions :: throwIfFatal detected a jvm fatal exception, which is thrown and logged below:
java.lang.NoClassDefFoundError: org/apache/hc/core5/reactive/ReactiveEntityProducer
	at org.springframework.http.client.reactive.HttpComponentsClientHttpRequest.toRequestProducer(HttpComponentsClientHttpRequest.java:172) ~[spring-web-5.3.27.jar:5.3.27]
	at org.springframework.http.client.reactive.HttpComponentsClientHttpConnector.execute(HttpComponentsClientHttpConnector.java:119) ~[spring-web-5.3.27.jar:5.3.27]
	at org.springframework.http.client.reactive.HttpComponentsClientHttpConnector.lambda$connect$1(HttpComponentsClientHttpConnector.java:115) ~[spring-web-5.3.27.jar:5.3.27]
	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44) ~[reactor-core-3.4.29.jar:3.4.29]`

cah292 commented:
@jhulslander You may need to add the `httpcore5-reactive` dependency to the POM; that's in KualiCo's latest POM but not in the 2022-10-19 patch. I also saw that KualiCo is pulling in a newer version of the `httpcore5-h2` transitive dependency; I'm not certain whether that extra step is needed.

cah292 commented:
@jhulslander If implementing your RESTful client calls via WebClient ends up being impractical on the current version of the financials code, then you could create a follow-up user story to perform that conversion once we upgrade to the appropriate financials release.

jhulslander commented:
as requested, I created KFSPTS-29123 to add unit tests for the webservice calls and updated KFSPTS-28830 to include this new code

jhulslander commented:
@cah292 this is ready for review again.  I have reverted back to older webservicd libraries.  Please see my notes about the JIra I created to add in unit tests, and I added a note to the webservice library about this new code

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -141,6 +156,22 @@ public <T> T unmarshalStream(InputStream inputStream, Class<T> clazz)
+
+        XMLInputFactory inputFactory = XMLInputFactory.newFactory();
+        inputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false);
+        XMLStreamReader streamReader = inputFactory.createXMLStreamReader(inputStream);
```
I would suggest specifying the charset when creating the stream reader (or create and use an intermediate `InputStreamReader` that specifies the charset). Sorry for not mentioning this earlier.

jhulslander replied:
inputFactory.createXMLStreamReader doesn't have an arg option for a character set, but it does have one that takes an encoding string, so I used the following and seems to work in a unit test, I will do an application test before I open this for review again
`streamReader = inputFactory.createXMLStreamReader(wrappedStream, StandardCharsets.UTF_8.name());`

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -141,6 +156,22 @@ public <T> T unmarshalStream(InputStream inputStream, Class<T> clazz)
+
+        Unmarshaller unmarshaller = jc.createUnmarshaller();
+        T response = (T) unmarshaller.unmarshal(streamReader);
+        CuXMLStreamUtils.closeQuietly(streamReader);
```
This close-quietly call should be placed in a `finally` block.

jhulslander replied:
ok

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/jaggaer/supplier/xml/SupplierSyncMessageCornellTest.java:
```diff
@@ -54,6 +67,40 @@ static Stream<Arguments> testSupplierSyncMessageArguments() {
+                + "    </SupplierResponseMessage>\n"
+                + "</SupplierSyncMessage>";
+        
+        SupplierSyncMessage synchMessage = marshalService.unmarshalStringIgnoreDtd(xmlString, SupplierSyncMessage.class);
```
Should this variable be renamed to `syncMessage`?

jhulslander replied:
ok

cah292 reviewed (changes requested):
Changes look good in general, though I found some areas that may need further cleanup.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,220 @@
+        results.setFileProcessedByJaggaer(true);
+
+        if (responseMessage.getStatus().getErrorMessages().isEmpty()) {
+            results.setErrorMessage(StringUtils.EMPTY);
```
Yes, that is intentional.  The goal was to have the last try be the results that are reported.  Either a success will happen or it will fail.  The report will note however it failed on the last try.  The job log should have details about each failure if a developer needs to review what happened.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,220 @@
+
+    private void generateResultsReport(List<JaggaerUploadFileResultsDTO> resultsList) {
+        /*
+         * A full report will be implemented on Jira KFSPTS-29084 + * Until then, this
```
will remove

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java:
```diff
@@ -1,42 +1,220 @@
+        boolean successfulCall = false;
+        int numberOfAttempts = 1;
+
+        while (!successfulCall && numberOfAttempts < getMaximumNumberOfRetries()) {
```
OK I will switch

jhulslander commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -121,6 +127,15 @@ public <T> T unmarshalString(String xmlString, Class<T> clazz) throws JAXBExcept
+    @Override
+    public <T> T unmarshalStringIgnoreDtd(String xmlString, Class<T> clazz)
+            throws JAXBException, XMLStreamException, IOException {
+        InputStream xmlStream = IOUtils.toInputStream(xmlString, StandardCharsets.UTF_8);
```
I moved InputStream xmlStream into the try()

jhulslander commented on src/main/java/edu/cornell/kfs/sys/service/impl/CUMarshalServiceImpl.java:
```diff
@@ -121,6 +127,15 @@ public <T> T unmarshalString(String xmlString, Class<T> clazz) throws JAXBExcept
+    public <T> T unmarshalStringIgnoreDtd(String xmlString, Class<T> clazz)
+            throws JAXBException, XMLStreamException, IOException {
+        InputStream xmlStream = IOUtils.toInputStream(xmlString, StandardCharsets.UTF_8);
+        try (CloseShieldInputStream wrappedStream = new CloseShieldInputStream(xmlStream);) {
```
ok

cah292 reviewed (approved):
Changes look good, unit tests pass, my concerns have been addressed and the related SQL changes are also ready. Merging.
